### PR TITLE
new updateCheck API with auth

### PR DESCRIPTION
**Pull request by scrajesh**
- new `/auth/updateCheck` api requires SSO auth token to get authorized
- this new API if authenticated, it supplies updates of any package as per the incoming request
- the existing `/updateCheck` supply updates only to those packages that are public
- Packages those  required authentication to get an upgrade can be protected through CCM configs `packsProtected`
- above CCM config will have the deployment keys of the packages to be protected

### CODE REVIEW

File: electrode-ota-server-routes-acquisition/src/is-protected.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+// check if incoming deployment key is in the protected packages list
+const isProtected = (key, protectedPacks) => {
+    if (key && protectedPacks && typeof protectedPacks === "string") {
+        const packs = protectedPacks.split(",");
```
[the remark this answers is not in the document]
Comment (scrajesh) [reply electrode-ota-server-routes-acquisition/src/is-protected.js:5]: Deployment keys are case-sensitive, added unit tests for the same.

File: electrode-ota-server-routes-acquisition/src/handlers.js [hunk 2]
```diff
@@ -17,6 +19,15 @@ const handles = {
         const snakeCaseRequested = /update_check/.test(action);
         // if snake_case request? convert to camelCase
         const qs = snakeCaseRequested ? keysToCamelOrSnake(request.query) : request.query;
+        // retrieve the protected packages list
+        const packsProtected = ccm("packsProtected");
+        const isProtectedPack = isProtected(qs.deploymentKey, packsProtected);
```
[the remark this answers is not in the document]
Comment (scrajesh) [reply electrode-ota-server-routes-acquisition/src/handlers.js:24]: added unit tests for isProtected()

**Comment by scrajesh** [comment]
> no tests

added unit tests for isProtected()

**Comment by scrajesh** [comment]
> who's calling /auth/updateCheck ? are they creating a custom ota client?
> where's the sso part of this code?

**Comment by scrajesh** [comment]
> So this change basically says, if updateCheck is requested at route `/auth/updateCheck` for a package that's been listed in CCM as protected, then return 401.
> 
> So how does a protected package get authed and goes through? I don't see any code for doing auth work. Is that already existing or something?

Actually if /updateCheck is trying to access a protected package, it issues 401.
Auth is implemented in Walmart's instance of the OTA Server.

**Comment by scrajesh** [comment]
> who's calling /auth/updateCheck ? are they creating a custom ota client?
> where's the sso part of this code?

Yes, it has to be a custom OTA client(will explore more on this).
SSO part can be found in Walmart's instance of the OTA Server.

File: electrode-ota-server-routes-acquisition/test/is-protected-test.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    it("should return true if deploymentKey is protected", () => {
+        const protectedPacks = "foo-bar-eeWWW-lorem-IpsiUM-FoO-bZr,foo-bar-eeWWW-lorem-IpsiUM-FoO-bAr,foo-bar-blah-blah-meh";
+        expect(isProtected(dKey, protectedPacks)).to.be.equal(true);
+    });
+
+    it("deploymentKey should be case-sensitive", () => {
+        let protectedPacks = "foo-bar,blah-blah-meh-blah,f0o-Blah-bAr-m3h";
+        expect(isProtected("Foo-Bar", protectedPacks)).to.be.equal(false);
+        expect(isProtected("f0o-Blah-bAr-m3h", protectedPacks)).to.be.equal(true);
+    });
```
Comment (datvong-wm) [inline electrode-ota-server-routes-acquisition/test/is-protected-test.js:35]: add another test for `"foo-bar, blah-blah-meh-blah, f0o-Blah-bAr-m3h"`

File: electrode-ota-server-routes-acquisition/test/update-check-test.js [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    it("should return a valid response for /updateCheck request", () => {
+        ccm = () => "";
+        reply = result => {
+            expect(result.updateInfo).to.equal(updateInfo);
+        };
+        handlers.updateCheck(request, reply, logger, ccm, acquisitionUpdateCheck);
+    });
+
+    it("should return a valid response for /auth/updateCheck request", () => {
+        request.url.pathname = "auth/updateCheck";
```
Comment (datvong-wm) [inline electrode-ota-server-routes-acquisition/test/update-check-test.js:66]: why do u have this?

File: electrode-ota-server-routes-acquisition/src/handlers.js [hunk 5]
```diff
@@ -17,6 +19,18 @@ const handles = {
         const snakeCaseRequested = /update_check/.test(action);
         // if snake_case request? convert to camelCase
         const qs = snakeCaseRequested ? keysToCamelOrSnake(request.query) : request.query;
+        // validate query-parameters
+        missingParameter(qs.deploymentKey, `Deployment key missing`);
+        missingParameter(qs.appVersion, `appVersion missing`);
```
Comment (datvong-wm) [inline electrode-ota-server-routes-acquisition/src/handlers.js:24]: do u know if there are any existing /updateCheck calls that has missing appVersion/deploymentKey?

File: electrode-ota-server-routes-acquisition/test/is-protected-test.js (hunk 3 above)
Comment (scrajesh) [reply electrode-ota-server-routes-acquisition/test/is-protected-test.js:37]: added

File: electrode-ota-server-routes-acquisition/test/update-check-test.js (hunk 4 above)
Comment (scrajesh) [reply electrode-ota-server-routes-acquisition/test/update-check-test.js:66]: the request object is one of the param to this method, hence need to set the path info accordingly

File: electrode-ota-server-routes-acquisition/src/handlers.js (hunk 5 above)
Comment (scrajesh) [reply electrode-ota-server-routes-acquisition/src/handlers.js]: wee already have a check here: https://github.com/scrajesh/electrode-ota-server/blob/master/electrode-ota-server-model-acquisition/src/acquisition.js#L30

It's just moved a bit up.

File: electrode-ota-server-model-acquisition/src/acquisition.js [hunk 6]
```diff
@@ -27,7 +27,7 @@ export default (options, dao, weighted, _download, manifest, logger) => {
          */
         updateCheck(params) {
 
-            missingParameter(params.deploymentKey, `Deployment key missing`);
+            //missingParameter(params.deploymentKey, `Deployment key missing`);
```
Comment (datvong-wm) [inline electrode-ota-server-model-acquisition/src/acquisition.js:30]: why are we commenting this out?

File: electrode-ota-server-model-acquisition/src/acquisition.js [hunk 7]
```diff
@@ -27,7 +27,7 @@ export default (options, dao, weighted, _download, manifest, logger) => {
...
             missingParameter(params.appVersion, `appVersion missing`);
```
Comment (datvong-wm) [inline electrode-ota-server-model-acquisition/src/acquisition.js:31]: Let's move appVersion to the other location as well.

File: electrode-ota-server-routes-acquisition/test/is-protected-test.js (hunk 3 above)
Comment (datvong-wm) [reply electrode-ota-server-routes-acquisition/test/is-protected-test.js:37]: Add
` expect(isProtected("foo-bar", "foo-bar, blah-blah-meh-blah, f0o-Blah-bAr-m3h")).to.be.equal(true);`

File: electrode-ota-server-routes-acquisition/test/update-check-test.js (hunk 4 above)
Comment (datvong-wm) [reply electrode-ota-server-routes-acquisition/test/update-check-test.js:66]: ic.  should it be  "auth/updateCheck" or "/auth/updateCheck"?

File: electrode-ota-server-routes-acquisition/test/update-check-test.js [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    });
+
+    it("should throw Deployment key missing", () => {
+        request.query.appVersion = null;
+        expect(handlers.updateCheck.bind(handlers, request, reply, logger, ccm, acquisitionUpdateCheck)).to.throw("appVersion missing");
+    });
+
+    it("should throw Unauthorized for /updateCheck asking for a protected package", () => {
+        expect(handlers.updateCheck.bind(handlers, request, reply, logger, ccm, acquisitionUpdateCheck)).to.throw("Unauthorized");
+    });
```
Comment (datvong-wm) [inline electrode-ota-server-routes-acquisition/test/update-check-test.js:55]: Why is this test throwing Unauthorized when your url is /updateCheck?

Comment (datvong-wm) [reply electrode-ota-server-routes-acquisition/test/update-check-test.js:55]: Can you add tests for "/auth/updateCheck" and "/auth/update_check"

File: electrode-ota-server-model-acquisition/src/acquisition.js (hunk 7 above)
Comment (scrajesh) [reply electrode-ota-server-model-acquisition/src/acquisition.js]: moved both up to the route handler, as of not this method is invoked from route handler.

File: electrode-ota-server-routes-acquisition/test/update-check-test.js (hunk 8 above)
Comment (scrajesh) [reply electrode-ota-server-routes-acquisition/test/update-check-test.js:56]: added more tests

File: electrode-ota-server-routes-acquisition/test/update-check-test.js (hunk 4 above)
Comment (scrajesh) [reply electrode-ota-server-routes-acquisition/test/update-check-test.js:92]: the pathname doesn't have `/`, verified in Staging.

File: electrode-ota-server-routes-acquisition/test/is-protected-test.js (hunk 3 above)
Comment (scrajesh) [reply electrode-ota-server-routes-acquisition/test/is-protected-test.js:40]: added the same

Comment (datvong-wm) [reply electrode-ota-server-routes-acquisition/test/is-protected-test.js:40]: where did u add
expect(isProtected("foo-bar", "foo-bar, blah-blah-meh-blah, f0o-Blah-bAr-m3h")).to.be.equal(true);

Comment (scrajesh) [reply electrode-ota-server-routes-acquisition/test/is-protected-test.js:40]: Added here: https://github.com/scrajesh/electrode-ota-server/blob/new-api/electrode-ota-server-routes-acquisition/test/is-protected-test.js#L38

File: electrode-ota-server-routes-acquisition/src/handlers.js [hunk 9]
```diff
@@ -17,6 +18,18 @@ const handles = {
...
+        // retrieve the protected packages list
+        const packsProtected = ccm("packsProtected");
+        const isProtectedPack = isProtected(qs.deploymentKey, packsProtected);
+        // check if the request is through authenticated api?
+        //  if so, check if the requested package is protected?
+        if (/^updateCheck|update_check/.test(action) && isProtectedPack) {
```
Comment (datvong-wm) [inline electrode-ota-server-routes-acquisition/src/handlers.js:29]: this regex is wrong

Comment (scrajesh) [reply electrode-ota-server-routes-acquisition/src/handlers.js]: modified the regex

**Review by datvong-wm — APPROVED** [review]

**Pull request merged**